packages/gatsby-plugin-sharp/src/__tests__/utils.js

```diff
@@ -39,3 +40,190 @@ describe(`createGatsbyProgressOrFallbackToExternalProgressBar`, () => {
+  })
+})
+
+describe(`calculateImageSizes (fluid & constrained)`, () => {
+  it(`should throw if maxWidth is less than 1`, () => {
+    const args = {
+      layout: `fluid`,
+      maxWidth: 0,
+      file,
+      imgDimensions,
+    }
+    const getSizes = () => calculateImageSizes(args)
+    expect(getSizes).toThrow()
+  })
+
+  it(`should include the original size of the image when maxWidth is passed`, () => {
+    const args = {
+      layout: `fluid`,
+      maxWidth: 400,
+      file,
+      imgDimensions,
+    }
+    const sizes = calculateImageSizes(args)
+    expect(sizes).toContain(400)
+  })
+
+  it(`should include the original size of the image when maxWidth is passed for a constrained image`, () => {
+    const args = {
+      layout: `constrained`,
+      maxWidth: 400,
+      file,
+      imgDimensions,
+    }
+    const sizes = calculateImageSizes(args)
+    expect(sizes).toContain(400)
+  })
+
+  it(`should include the original size of the image when maxHeight is passed`, () => {
+    const args = {
+      layout: `fluid`,
+      maxHeight: 300,
+      file,
+      imgDimensions,
+    }
+    const sizes = calculateImageSizes(args)
+    expect(sizes).toContain(450)
+  })
+
+  it(`should create images of different sizes (0.25x, 0.5x, 1x, 2x, and 3x) from a maxWidth`, () => {
+    const args = {
+      layout: `fluid`,
+      maxWidth: 320,
+      file,
+      imgDimensions,
+    }
+    const sizes = calculateImageSizes(args)
+    expect(sizes).toEqual(expect.arrayContaining([80, 160, 320, 640, 960]))
+  })
+
+  it(`should create images of different sizes (0.25x, 0.5x, 1x) without any defined size`, () => {
+    const args = {
+      layout: `fluid`,
+      file,
+      imgDimensions,
+    }
+    const sizes = calculateImageSizes(args)
+    expect(sizes).toEqual(expect.arrayContaining([200, 400, 800]))
+  })
+
+  it(`should return sizes of provided srcSetBreakpoints`, () => {
+    const srcSetBreakpoints = [50, 70, 150, 250, 300]
+    const maxWidth = 500
+    const args = {
+      layout: `fluid`,
+      maxWidth,
+      srcSetBreakpoints,
+      file,
+      imgDimensions,
+    }
+
+    const sizes = calculateImageSizes(args)
+    expect(sizes).toEqual(expect.arrayContaining([50, 70, 150, 250, 300, 500]))
+  })
+
+  it(`should reject any srcSetBreakpoints larger than the original width`, () => {
+    const srcSetBreakpoints = [
+      50,
+      70,
+      150,
+      250,
+      1250, // shouldn't be included, larger than original width
+    ]
+    const maxWidth = 1500 // also shouldn't be included
+    const args = {
+      layout: `fluid`,
+      maxWidth,
+      srcSetBreakpoints,
+      file,
+      imgDimensions,
+    }
+
+    const sizes = calculateImageSizes(args)
+    expect(sizes).toEqual(expect.arrayContaining([50, 70, 150, 250]))
+    expect(sizes).toEqual(expect.not.arrayContaining([1250, 1500]))
+  })
+
+  it(`should also include sizes from srcSetBreakpoints when outputPixelDensities are also passed in`, () => {
```

Comment:
I think if we're passing in srcsetBreakpoints we should _only_ use those

Reply:
(Related to https://github.com/gatsbyjs/gatsby/pull/27554#discussion_r507853223)

Reply:
Updated this, but a second set of eyes on the logic would be much appreciated 😄